File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/python/greedy_algorithms/algorithms/job_scheduling.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Job:
    name: str
    deadline: int
    profit: int


def schedule_jobs(jobs: list[Job]) -> dict[str, object]:
    if any(job.deadline <= 0 for job in jobs):
        raise ValueError("job deadlines must be positive")

    max_deadline = max((job.deadline for job in jobs), default=0)
    slots: list[Job | None] = [None] * max_deadline

    for job in sorted(jobs, key=lambda item: item.profit, reverse=True):
        for slot in range(min(job.deadline, max_deadline) - 1, -1, -1):
            if slots[slot] is None:
                slots[slot] = job
                break

    selected = [job for job in slots if job is not None]
    return {
        "jobs": selected,
        "total_profit": sum(job.profit for job in selected),
    }

File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/java/src/main/java/com/example/greedy/algorithms/JobScheduling.java
package com.example.greedy.algorithms;

import com.example.greedy.model.Job;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public final class JobScheduling {
    private JobScheduling() {
    }

    public static final class Result {
        private final List<Job> jobs;
        private final int totalProfit;

        public Result(List<Job> jobs, int totalProfit) {
            this.jobs = List.copyOf(jobs);
            this.totalProfit = totalProfit;
        }

        @Override
        public String toString() {
            return "Result{jobs=" + jobs + ", totalProfit=" + totalProfit + "}";
        }
    }

    public static Result schedule(List<Job> jobs) {
        int maxDeadline = 0;
        for (Job job : jobs) {
            if (job.deadline() <= 0) {
                throw new IllegalArgumentException("job deadlines must be positive");
            }
            maxDeadline = Math.max(maxDeadline, job.deadline());
        }

        Job[] slots = new Job[maxDeadline];
        for (Job job : jobs.stream()
                .sorted(Comparator.comparingInt(Job::profit).reversed())
                .collect(Collectors.toList())) {
            for (int slot = Math.min(job.deadline(), maxDeadline) - 1; slot >= 0; slot--) {
                if (slots[slot] == null) {
                    slots[slot] = job;
                    break;
                }
            }
        }

        List<Job> selected = new ArrayList<>();
        int totalProfit = 0;
        for (Job job : slots) {
            if (job != null) {
                selected.add(job);
                totalProfit += job.profit();
            }
        }
        return new Result(selected, totalProfit);
    }
}
